internal/uvm/create.go

```diff
@@ -286,23 +315,46 @@ func Create(opts *UVMOptions) (*UtilityVM, error) {
 		}
 
 		if opts.ConsolePipe != "" {
-			hcsDocument.VirtualMachine.Chipset.UEFI.BootThis.OptionalData += " console=ttyS0,115200"
-			hcsDocument.VirtualMachine.Devices.COMPorts = &schema2.VirtualMachinesResourcesComPortsV2{Port1: opts.ConsolePipe}
+			vmDebugging = true
+			kernelArgs += " console=ttyS0,115200"
+			vm.Devices.COMPorts = &schema2.VirtualMachinesResourcesComPortsV2{Port1: opts.ConsolePipe}
 		}
 
 		if opts.EnableGraphicsConsole {
-			hcsDocument.VirtualMachine.Chipset.UEFI.BootThis.OptionalData += " console=tty"
-			hcsDocument.VirtualMachine.Devices.Keyboard = &schema2.VirtualMachinesResourcesKeyboardV2{}
-			hcsDocument.VirtualMachine.Devices.Rdp = &schema2.VirtualMachinesResourcesRdpV2{}
-			hcsDocument.VirtualMachine.Devices.VideoMonitor = &schema2.VirtualMachinesResourcesVideoMonitorV2{}
+			vmDebugging = true
+			kernelArgs += " console=tty"
+			vm.Devices.Keyboard = &schema2.VirtualMachinesResourcesKeyboardV2{}
+			vm.Devices.Rdp = &schema2.VirtualMachinesResourcesRdpV2{}
+			vm.Devices.VideoMonitor = &schema2.VirtualMachinesResourcesVideoMonitorV2{}
+		}
+
+		if !vmDebugging {
+			// Terminate the VM if there is a kernel panic.
+			kernelArgs += " panic=-1"
 		}
 
 		if opts.KernelBootOptions != "" {
-			hcsDocument.VirtualMachine.Chipset.UEFI.BootThis.OptionalData = hcsDocument.VirtualMachine.Chipset.UEFI.BootThis.OptionalData + fmt.Sprintf(" %s", opts.KernelBootOptions)
+			kernelArgs += " " + opts.KernelBootOptions
+		}
+
+		// Start GCS with stderr pointing to the vsock port created below in
+		// order to forward guest logs to logrus.
+		initArgs := fmt.Sprintf("/bin/vsockexec -e %d /bin/gcs -log-format json -loglevel %s",
+			linuxLogVsockPort,
+			logrus.StandardLogger().Level.String())
+
+		if vmDebugging {
+			// Launch a shell on the console.
+			initArgs = `sh -c "` + initArgs + ` & exec sh"`
```

Comment:
Will this break hcsdiag console -uvm for the normal case? (I assume it relies on init starting sh prior to the other changes you made).

Reply:
hcsdiag console -uvm sends a command to hcs to start another sh, so it's fine.